packages/pagination/src/Pagination.js

```diff
@@ -0,0 +1,97 @@
+import React, { useState, useContext } from 'react';
+import PropTypes from 'prop-types';
+import isFunction from 'lodash.isfunction';
+import { useEffectAsync, useToggle } from '@availity/hooks';
+
+export const PaginationContext = React.createContext();
+
+export const usePagination = () => useContext(PaginationContext);
+
+// todos
+// Add another `useEffect` for only updating the items without calling the function if the `itemsPerPage` prop changes
+
+/**
+ * Pagination Provider that takes in an async function or items and provides a list of items
+ * @param {Function|Object} items - The Items or function to use for pagination
+ * @param {number} itemsPerPage - The items to show per page
+ */
+const Pagination = ({ items: theItems, itemsPerPage, children }) => {
+  const [currentPage, setPage] = useState(1);
+  const [pageData, setPageData] = useState({
+    total: theItems != null ? theItems.totalCount : 0,
+    pages: [],
+    page: {},
+    lower: 0,
+    upper: 0,
+  });
+
+  const [loading, toggleLoading] = useToggle(true);
+
+  useEffectAsync(async () => {
+    if (!loading) {
+      toggleLoading();
+    }
+
+    // If the items is a function then await the resposne in case of async actions
+    const { items, totalCount } = isFunction(theItems)
+      ? await theItems(currentPage, itemsPerPage)
+      : { items: theItems };
+
+    // Get index of item at the start of the currentPage
+    const lower = currentPage === 1 ? 1 : (currentPage - 1) * itemsPerPage + 1;
+
+    // Get the index of the item at the cut-off of the itemPerPage Count
+    const upper =
+      items.length - currentPage * itemsPerPage > 0
+        ? itemsPerPage * currentPage
+        : items.length;
+
+    // Slice that data if it was given from a function since we don't know if its returning all or not
+    // todo - add prop if needed to handle this
+    const pageData = isFunction(theItems)
+      ? items
+      : items.slice(lower - 1, upper);
+
+    // Get page numbers
+    const pages = Array.from(
+      { length: Math.ceil((totalCount || items.length) / itemsPerPage) },
+      (v, k) => k + 1
+    );
+
+    setPageData({
+      total: totalCount || items.length,
+      pages,
+      page: { number: currentPage, items: pageData },
+      lower,
+      upper,
+    });
+
+    toggleLoading();
```

Comment:
Since this function is potentially async, this statement may get triggered when loading has already been toggled off. You may want to support passing a param to the toggle function to explicitly set the value (like some other hook library toggles support) or wrap this in a condition to only run if loading is true.

Reply:
Similar to this comment above where multiple requests are in flight (such as the person hitting the next page twice before the first request finishes or just clicking page numbers before each request finishes), how does this handle responses out of order (such as the first request finishing after the second request)?
Usually this causes the wrong page to display. Say the user clicks page 2 and then page 3. The page 3 request finishes and displays the page 3 results, but then the page 2 request finishes... now `currentPage` is still 3, but `pageData` is the items for page 2.
You may want to cancel pending requests before making new ones or just keep track of the most recent request and only care about its response.

Reply:
I just added in the latest commit the `localstorage-core` to handle the current page and keep the component in a loading state until the latest page is rendered. We can potentially add the fetch abort options into this. I have seen some people creating a hooks that handle all of that for you and it would be interesting to see the `api-axios` resource in something like the below down the road.

 ```javascript
const [resource,getResource] = useResource(AvOrganizationResource)`
``` 

Inspiration: https://github.com/schettino/react-request-hook

Reply:
I don't think you need the localstorage stuff. Firstly, it's global (at least all of the pagination components will share the same store and thus the same `current-page`). Not a big deal if there is only ever 1 pagination component rendered at a time.
Second, I believe the values you end up comparing will always match. `currentPage` is scoped out-side of the `useEffectAsync` function and will always be the most recent page. `avLocalStorage.set('current-page', currentPage);` is the first thing to run, so in the most recent/latest time it runs, it will set the value in local storage to the most recent value, which as mentioned before will be the same value used when comparing the older responses in `if (!isEqual(avLocalStorage.get('current-page'), currentPage))`
Since `currentPage` is a primitive, you can create a copy of `currentPage` within the `useEffectAsync` function (something like `const thisPage = currentPage`) and then when `currentPage` changes and is referenced/accessed later it will be the most recent page and `thisPage` will be the page at the time the `useEffectAsync` function ran.

Reply:
Due to how to the digest cycle works for hooks. Referencing a variable that is called later will actually not take the latest value and instead will just use the current value at the time of the update.

Would it be better then to add a constructor onto the `localStorage` object and optional pass in a prefix or hash for that particular instance. This way the items used in that local storage instance are always guaranteed to be unique?

Reply:
I was mostly talking about variable scoping in javascript, but if hooks do something which make javascript not lexically scoped then that's cool.